### PR TITLE
KAFKA-9140: Also reset join future when generation was reset in order to re-join

### PR DESCRIPTION
**Pull request by guozhangwang**
Otherwise the join-group would not be resend and we'd just fall into the endless loop.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
cc @ableegoldman

**Comment by ableegoldman** [comment]
Nice find! Kicked off the version probing system test on this branch: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3409

**Review by abbccdda — COMMENTED** [review]
Could we comment the functionality of `resetStateAndRejoin` vs `resetJoinGroupFuture` or consider renaming these two functions so that we don't fall into the assumption that calling `resetStateAndRejoin` is not doing what caller expects?

**Review by ableegoldman — APPROVED** [review]
System tests passed 20/20 runs, LGTM (+1 Boyang's suggestion about adding comments though)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 1]
```diff
@@ -426,8 +426,10 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     needsJoinPrepare = true;
                 } else {
                     log.info("Generation data was cleared by heartbeat thread. Initiating rejoin.");
-                    resetStateAndRejoin();
-
+                    synchronized (AbstractCoordinator.this) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Resetting state should be protected with synchronization.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java [hunk 2]
```diff
@@ -2345,12 +2345,22 @@ public void testConsumerRejoinAfterRebalance() throws Exception {
...
-            //Join future should succeed but generation already cleared so result of join is false.
+            // Join future should succeed but generation already cleared so result of join is false.
             res = coordinator.joinGroupIfNeeded(time.timer(1));
 
             assertFalse(res);
             assertFalse(client.hasPendingResponses());
             assertFalse(client.hasInFlightRequests());
+
+            // Retry join should then succeed
+            client.prepareResponse(joinGroupFollowerResponse(generationId, memberId, "leader", Errors.NONE));
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:2356]: Augmented this unit test, without the fix above this test will now fail.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 3]
```diff
@@ -447,10 +449,12 @@ else if (!future.isRetriable())
         return true;
     }
 
+    // reset join-group future in order to retry the join-group request
     private synchronized void resetJoinGroupFuture() {
         this.joinFuture = null;
     }
 
+    // reset state to UNJOINED and request re-join
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Do these comments have much value? They basically just describe the two lines of code below.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Why don't we make `resetStateAndRejoin` synchronized?

Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: In other two callers of `resetStateAndRejoin` it is already synchronized at the caller, so I just follow the same pattern here.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 3 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: This is commented by @abbccdda who are trying to clarify the different functionalities here, after second thought I think it makes more sense to just comment on the original call for different cases.

**Comment by guozhangwang** [comment]
@hachikuji addressed your comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Yes, but that is why this was a bug, right?

Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: The original bug was on `generation`, but I think we should also be careful on `state` as well.  I thought doing the synchronization on the caller is fine but maybe just enforce it at the callee is better for future-proof, changing it then.

**Review by hachikuji — APPROVED** [review]
LGTM

**Comment by guozhangwang** [comment]
Cherry-picked to 2.4.

**Pull request merged**